airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java

```diff
@@ -79,15 +93,74 @@ public void start() throws Exception {
   }
 
   @Override
-  public void accept(final AirbyteMessage message) throws Exception {
+  public void accept(final String messageString, final Integer sizeInBytes) throws Exception {
     Preconditions.checkState(hasStarted, "Cannot accept records until consumer has started");
+    propagateFlushWorkerExceptionIfPresent();
     /*
      * intentionally putting extractStream outside the buffer manager so that if in the future we want
-     * to try to use a threadpool to partial deserialize to get record type and stream name, we can do
-     * it without touching buffer manager.
+     * to try to use a thread pool to partially deserialize to get record type and stream name, we can
+     * do it without touching buffer manager.
      */
-    extractStream(message)
-        .ifPresent(streamDescriptor -> bufferEnqueue.addRecord(streamDescriptor, message));
+    deserializeAirbyteMessage(messageString)
+        .ifPresent(message -> {
+          if (message.getType() == Type.RECORD) {
+            validateRecord(message);
+          }
+
+          bufferEnqueue.addRecord(message, sizeInBytes);
+        });
+  }
+
+  private Optional<PartialAirbyteMessage> deserializeAirbyteMessage(final String messageString) {
+    final Optional<PartialAirbyteMessage> messageOptional = Jsons.tryDeserialize(messageString, PartialAirbyteMessage.class)
+        .map(partial -> partial.withSerialized(messageString));
+    if (messageOptional.isPresent()) {
+      return messageOptional;
+    } else {
+      if (isStateMessage(messageString)) {
```

Comment:
I'm likely missing something but this logic seems to prevent the `bufferEnqueue` from receiving a STATE message, no?